File: solutions/problem_0532_k_diff_pairs_in_an_array/solution.py
"""
Problem 532: K-diff Pairs in an Array
====================================
Difficulty: Medium
Tags: Array, Hash Table, Two Pointers, Binary Search, Sorting
Pattern: Hash Map

Time Complexity:  O(n)
Space Complexity: O(n)
"""

class Solution:
    def findPairs(self, nums: list[int], k: int) -> int:
        from collections import Counter
        counter = Counter(nums)
        result = 0
        for num in counter:
            if k == 0 and counter[num] > 1:
                result += 1
            elif k > 0 and num + k in counter:
                result += 1
        return result

File: solutions/problem_0532_k_diff_pairs_in_an_array/solution.cpp
/*
 * Problem 532: K-diff Pairs in an Array
 * ====================================
 * Difficulty: Medium
 * Tags: Array, Hash Table, Two Pointers, Binary Search, Sorting
 * Pattern: Hash Map
 *
 * Time Complexity:  O(n)
 * Space Complexity: O(n)
 */

#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        // Binary search - O(log n) time, O(1) space
        int lo = 0, hi = nums.size() - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] == k) {
                return mid;
            } else if (nums[mid] < k) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        return 0;
    }
};
